docs(refactor): refactor and refresh of filter content

PaulRMellor opened the pull request:
**Documentation**

Refactor and refresh of the docs.
Updates include the following:

- Refactors the files into modular format
- Introduces a readme to decribe the guide and structure
- Global attributes file
- Splits filter content and KMS set up sections into separate chapters (assemblies)
- Rewrites for reuse and to clean up text

Scope of content updates include _Overview and Record Encryption filter only_.
Cosmetic or minor content updates to other sections (which will be refreshed in subsequent PRs)

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Review performance test results. Ensure that any degradations to performance numbers are understood and justified.
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [x] Update documentation
- [ ] Reference relevant issue(s) and close them after merging
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

k-wall commented on docs/_assets/attributes.adoc:
```diff
@@ -0,0 +1,38 @@
+:start-script: https://github.com/kroxylicious/kroxylicious/blob/main/kroxylicious-app/src/assembly/kroxylicious-start.sh
+
+//Kafka links
+:ApacheKafkaSite: https://kafka.apache.org[Apache Kafka web site^]
```
I've become accustomed to seeing website spelt as a single word.  Was this intention?

k-wall commented on docs/assemblies/assembly-aws-kms.adoc:
```diff
@@ -0,0 +1,14 @@
+// file included in the following:
```
KMS currently belongs to exclusively to the record-encryption filter, so to see it further up the tree looks surprising.

k-wall replied:
I agree that KMSs will probably end up as a concept in Kroxylicious that is used by n filters.  On reflection, I think the path structure suggest by Paul is okay.

k-wall commented on docs/assemblies/assembly-record-encryption-filter.adoc:
```diff
@@ -0,0 +1,29 @@
+= Record Encryption filter
+
+[role="_abstract"]
+Streams for Apache Kafka Proxy's Record Encryption filter enhances the security of Kafka messages.
```
```suggestion
Kroxylicious's Record Encryption filter enhances the security of Kafka messages.
```

SamBarker replied:
Should we be extracting a variable so that `Kroxylicious` can be replaced downstream?

k-wall commented on docs/assemblies/assembly-record-encryption-filter.adoc:
```diff
@@ -0,0 +1,29 @@
+3. Configuring the filter within Kroxylicious.
+
+The filter integrates with a Key Management Service (KMS), which has ultimate responsibility for the safe storage of key material.
+Currently, the filter integrates with HashiCorp Vault and AWS Key Management Service as its KMS, though further supported KMS integrations are planned.
```
```suggestion
Currently, the filter integrates with either HashiCorp Vault or AWS Key Management Service as its KMS. Further KMS integrations are planned.
```

SamBarker replied:
```suggestion
The filter depends on a KMS implementation, currently HashiCorp Vault and AWS Key Management Service are supported. Users are able to supply implementations for their specific KMS systems. Support for further KMS systems will be based on demand.
```

k-wall commented on docs/modules/con-community-filters.adoc:
```diff
@@ -0,0 +1,10 @@
+
+[role="_abstract"]
+Community contributed filters are showcased in the
+https://ghub.com/kroxylicious/kroxylicious-community-gallery[Community Gallery].
```
```suggestion
https://github.com/kroxylicious/kroxylicious-community-gallery[Community Gallery].
```

k-wall commented:
@PaulRMellor thanks for getting this work underway.  I'm part way through my review. I'll continue tomorrow.

SamBarker commented on docs/assemblies/assembly-record-encryption-filter.adoc:
```diff
@@ -0,0 +1,29 @@
+[role="_abstract"]
+Streams for Apache Kafka Proxy's Record Encryption filter enhances the security of Kafka messages.
+The filter uses industry-standard cryptographic techniques to apply encryption to Kafka messages, ensuring the confidentiality of data stored in the Kafka Cluster.
+Streams for Apache Kafka Proxy centralizes topic-level encryption, ensuring streamlined encryption across Kafka clusters.
```
```suggestion
Kroxylicious centralizes topic-level encryption, ensuring streamlined encryption across Kafka clusters.
```

SamBarker commented on docs/assemblies/assembly-record-encryption-filter.adoc:
```diff
@@ -0,0 +1,29 @@
+[role="_abstract"]
+Streams for Apache Kafka Proxy's Record Encryption filter enhances the security of Kafka messages.
+The filter uses industry-standard cryptographic techniques to apply encryption to Kafka messages, ensuring the confidentiality of data stored in the Kafka Cluster.
+Streams for Apache Kafka Proxy centralizes topic-level encryption, ensuring streamlined encryption across Kafka clusters.
```
`centralizes` or `centralises` ?

k-wall replied:
Our DEV_GUIDE should call out whether the project is favo(u)ring the orthography of US or standard English.  We should then use it consistently across code and docs.

tombentley replied:
We should use US English for docs and the website, because (I _think_, maybe @PaulRMellor can confirm?) that's the standard used in other Red Hat docs and to which we will be held.

SamBarker replied:
_theatrical resigned sigh_

fine.

SamBarker commented on docs/assemblies/assembly-record-encryption-filter.adoc:
```diff
@@ -0,0 +1,29 @@
+2. Establishing the encryption keys within the KMS that will be used to encrypt the topics.
+3. Configuring the filter within Kroxylicious.
+
+The filter integrates with a Key Management Service (KMS), which has ultimate responsibility for the safe storage of key material.
```
```suggestion
The filter integrates with a Key Management Service (KMS), which has ultimate responsibility for the safe storage of sensitive key material.
```

?

or perhaps more extensively
```suggestion
The filter integrates with a Key Management Service (KMS), which has ultimate responsibility for managing the life cycle of sensitive key material.
```

SamBarker commented on docs/assemblies/assembly-record-encryption-filter.adoc:
```diff
@@ -0,0 +1,29 @@
+
+//overview of the record encryption process
+include::../modules/record-encryption/con-record-encryption-overview.adoc[leveloffset=+1]
+//setting up hashicorp vault
+include::assembly-hashicorp-vault.adoc[leveloffset=+1]
+//setting up AWS KMS
+include::assembly-aws-kms.adoc[leveloffset=+1]
```
We really need to be clear here that this is an `X or Y` choice! 

Should we maybe have `Set up a KMS` at `levelOffset=+1` and then the includes at `levelOffset+=2` ?

robobario replied:
I think it should be shuffled so that we have the generic configuration of the Record Encryption filter at the top level with KMS selections within like:

```
= Setting up the Record Encryption Filter
== Setting up a KMS
=== HashiCorp Vault KMS
=== AWS KMS
```
Though I wonder if we need the `Setting up a KMS` block, or can just say in the `Setting up the Record Encryption Filter` section that you need to pick one only.

SamBarker commented on docs/_assets/attributes.adoc:
```diff
@@ -0,0 +1,38 @@
+:numbered:
+:sectanchors!:
+:sectnums:
+:source-highlighter: highlight.js
```
This switches from `pygments` to `highlight.js` is that intentional / valid here upstream?

SamBarker commented on docs/modules/con-community-filters.adoc:
```diff
@@ -0,0 +1,10 @@
+= Community filters
+
+[role="_abstract"]
+Community contributed filters are showcased in the
```
Should we be saying something about support levels here? I.e. use at your own risk...

SamBarker commented on docs/modules/con-deploying.adoc:
```diff
@@ -1,13 +1,19 @@
 
+//== Providing implementations for facades
 
 == Configuring virtual clusters
```
We should probably add a note here to say users need one virtual cluster per listener per Kafka cluster.

SamBarker commented on docs/modules/con-operating.adoc:
```diff
@@ -1,6 +1,15 @@
+
 == Logging
 
 The Kroxylicious https://github.com/kroxylicious/kroxylicious/releases/latest[binary distributions] ship with https://logging.apache.org/log4j/2.x[log4j2] as a logging backend. To customize the logging configuration:
```
`latest` -> `:ProductVersion` ?

SamBarker commented on docs/modules/con-proxy-overview.adoc:
```diff
@@ -60,7 +59,7 @@ First let's define the concepts in the landscape surrounding Kroxylicious.
 Now let's define some concepts used within Kroxylicious itself.
 
-=== Virtual Cluster
+== Virtual cluster
```
Isn't `virtual cluster part of `How Kroxylicious works` ?

SamBarker commented on docs/modules/con-proxy-overview.adoc:
```diff
@@ -146,9 +145,7 @@ filter.
-
-
-=== Filter Chain
+=== Filter chain
```
Now that `Virtual Cluster` and `Target Cluster` only have `==` this makes `FilterChain` a sub point of `Target Cluster` which is wrong.

SamBarker commented on docs/modules/con-proxy-overview.adoc:
```diff
@@ -233,7 +230,7 @@ Proxying the access of one or more clients to a particular cluster/broker that m
 // TODO include a diagram
```
We must have generated an appropriate diagram somewhere...

SamBarker commented on docs/modules/con-proxy-overview.adoc:
```diff
@@ -249,11 +246,8 @@ _Transparent_ multi-tenancy means each tenant has the illusion of having their o
+For example:
 
-== More about filters
+* Single proxy instance
```
Would it be clearer to describe `Single Proxy instance` as `Sidecar` ? 

We also miss embedded from the list (though I think thats intentional)

SamBarker commented on docs/modules/record-encryption/aws-kms/con-aws-kms-key-creation.adoc:
```diff
@@ -1,12 +1,15 @@
 usage.  Multi-region keys are supported.  It is not possible to make use of keys from other AWS accountsfootnote:[https://github.com/kroxylicious/kroxylicious/issues/1217].
 
-Give the key an alias following the link:setup.adoc#_establish_an_aliasing_convention_for_keys_within_aws_kms[alias naming convention] established above.
+Give the key an alias as described in xref:con-aws-kms-setup-{context}[].
```
the empty squire brackets looks odd presumably we need some text there for the link text?

SamBarker commented on docs/modules/record-encryption/con-record-encryption-overview.adoc:
```diff
@@ -0,0 +1,95 @@
+
+. The filter intercepts produce requests from producing applications and encrypts the records.
+. The produce request is forwarded to the broker. 
+. The filter intercepts fetch responses from consuming applications and decrypts the records.
```
```suggestion
. The filter intercepts fetch responses from the broker and transforms it by decrypting the records.
```

SamBarker commented on docs/modules/record-encryption/con-record-encryption-overview.adoc:
```diff
@@ -0,0 +1,95 @@
+
+The process is as follows:
+
+. The filter intercepts produce requests from producing applications and encrypts the records.
```
```suggestion
. The filter intercepts produce requests from producing applications and transforms it by encrypting the records.
```

SamBarker commented on docs/modules/record-encryption/con-record-encryption-overview.adoc:
```diff
@@ -0,0 +1,95 @@
+. Filter selects a KEK to apply.
+. Requests the KMS to generate a DEK for the KEK.
+. Uses an encrypted DEK (DEK encrypted with the KEK) to encrypt the record.
+. Replaces the original record with a cipher record (encrypted record, encrypted DEK, and metadata).
```
Is it a `cipher record` or `enciphered record` ? I'm not sure what the correct/ commonly accepted term is.

k-wall replied:
In the design we use the term "ciphertext record" to describe the record content that gets forwarded to the broker.   I think that's the term we should favour in the user facing documentation.

https://github.com/kroxylicious/kroxylicious/blob/4535cc01e7de2face81ea9d65827988f6d1c3a8a/kroxylicious-filters/kroxylicious-record-encryption/doc/design.adoc?plain=1#L231

SamBarker replied:
I'm all for consistency so `ciphertext record` would be the path of least resistance but still sounds a bit odd to me. 

The more I think about it the harder time I have coming up with a phrase that works for me, so lets just be consistent unless @PaulRMellor has better idea.

SamBarker commented on docs/modules/record-encryption/con-record-encryption-overview.adoc:
```diff
@@ -0,0 +1,95 @@
+. Uses the decrypted DEK to decrypt the encrypted record.
+. Replaces the cipher record with a decrypted record.
+
+The filter uses an LRU (least recently used) strategy for decrypted records.
```
The filter DOES NOT cache decrypted records, It caches Decrypted DEKS!

SamBarker commented on docs/modules/record-encryption/con-record-encryption-overview.adoc:
```diff
@@ -0,0 +1,95 @@
+The KMS doesn't store DEKs; they are stored as part of the cipher record in the broker.
+
+WARNING: The KMS must be available during runtime.
+If the KMS is unavailable, production and consumption through the filter become impossible until KMS service is restored.
```
The DEK caches mitigate this somewhat. Though I'm not sure if we want to call that out.

SamBarker reviewed (commented):
Thanks @PaulRMellor. I think the new structure makes sense and I'm generally happy with the PR. 

I appreciate this is largely a refactoring but in reading the diff there were a few content changes that seemed worth highlighting so please accept my apologies for picking on things you didn't actually write in the first place!

k-wall commented on docs/_assets/attributes.adoc:
```diff
@@ -0,0 +1,38 @@
+:icons: font
+
+//Latest version
+:ProductVersion: 0.6
```
Indeed, we can probably give the responsibilty to https://github.com/kroxylicious/kroxylicious/blob/4535cc01e7de2face81ea9d65827988f6d1c3a8a/scripts/stage-release.sh#L103

k-wall commented on docs/README.md:
```diff
@@ -0,0 +1,50 @@
+ and it provides content to help you understand and set up Kroxylicious proxies.
+
+<!-- omit from toc -->
+## Table of Contents
```
Perhaps the `Rendering documentation` section from the dev guide should be moved to this file and the DEV_GUIDE should link to this README.

k-wall commented on docs/index.adoc:
```diff
@@ -1,16 +1,24 @@
+include::assemblies/assembly-overview.adoc[leveloffset=+1]
 
-include::available-filters.adoc[leveloffset=1]
+//built-in filters
```
I think the filters ought to be grouped together in their own section 'filters'.  It seems weird for 'record encryption' to be at the same level as 'operating proxies'

k-wall commented:
Thanks @PaulRMellor that's my first sweep through do. I'll review again once you've had change to address comments.

robobario commented on docs/modules/con-proxy-overview.adoc:
```diff
@@ -17,7 +16,7 @@ But there are numerous practical reasons that can prevent this, for example:
 For example using Kroxylicious proxy it's easier to experiment with alternative transport protocols, such as Quic, or operating system APIs, such as io_uring, because there is already support for this in Netty, the networking framework on which Kroxylicious is built.
 
-== How it works
+== How Kroxylicious works
```
This shouldn't be nested under `Why use proxies`:

suggest:

```
= Why use proxies?
= How Kroxylicious works
== Virtual cluster
== Target cluster
== Filter chain
== Filter composition
== Implementation
== Deployment topologies
```

robobario commented on docs/modules/record-encryption/con-record-encryption-overview.adoc:
```diff
@@ -0,0 +1,95 @@
+// assembly-record-encryption-filter.adoc
+
+[id='con-topic-encryption-overview-{context}']
+= Encryption process
```
Maybe we could align it with the other heading `How Kroxylicious works` like`How Record Encryption works`.

robobario commented on docs/modules/record-encryption/proc-configuring-record-encryption-filter.adoc:
```diff
@@ -0,0 +1,49 @@
+// assembly-record-encryption.adoc
+
+[id='proc-configuring-record-encryption-filter-{context}']
+= Setting up the Record Encryption filter
```
I think this section should come before the KMS configuration information and should probably contain the KMS config sections.

robobario commented on docs/modules/record-encryption/proc-configuring-record-encryption-filter.adoc:
```diff
@@ -0,0 +1,49 @@
+
+. Configure a `RecordEncryption` type filter.
++
+.Example Record Encryption filter configuration
```
Side note, I wonder if we should include a complete configuration example somewhere. We currently have the template and then examples for each KMS. Maybe it would be useful somewhere to glue them together, maybe in an Example Configurations section? Somewhere the user can directly copy a whole config and get something that nearly works if they change the KMS address.

(not for this refactor PR)

robobario commented on docs/modules/con-deploying.adoc:
```diff
@@ -1,13 +1,19 @@
-== Providing implementations for facades
-
+Before configuring your virtual clusters, ensure that the necessary filter plugins are available. 
+Place the required filter JAR files in a designated directory accessible by Kroxylicious. 
```
Somehow we should communicate which Filters are included in the distribution and don't need to be placed anywhere. I'm not sure if that goes here, or in a table up near the Filters, or in each Filter's subsection.

robobario commented on docs/modules/con-deploying.adoc:
Contains a placeholder that could be removed:

```
== Configuring proxy plugins
Filter level configuration
```

robobario commented on docs/modules/con-operating.adoc:
The headings `Use an Alternative Log4j configuration file` and `Change the Root Log level` are indented too far:

It looks like:
```
9.1. Logging
9.1.1. 1. Use an Alternative Log4j configuration file
9.1.2. 2. Change the Root Log level

robobario commented on docs/modules/con-deploying.adoc:
Maybe it should be changed so that `Upstream TLS` and `Downstream TLS` sections are nested under `Transport Layer Security (TLS)`

robobario commented on docs/modules/con-custom-filters.adoc:
```diff
@@ -1,11 +1,11 @@
 = Custom Filters
 
+[role="_abstract"]
 Custom filters can be written in the Java programming language.
```
We should add that we support Java 17 language level.

robobario reviewed (commented):
Thanks @PaulRMellor!

PaulRMellor commented:
Thanks for the review comments @SamBarker 
Updates: https://github.com/kroxylicious/kroxylicious/pull/1366/commits/63cf76690c7505c7bd690d02fdb9aecbda1713dc

PaulRMellor commented:
Thanks for the review comments @k-wall 
Updates: https://github.com/kroxylicious/kroxylicious/pull/1366/commits/63cf76690c7505c7bd690d02fdb9aecbda1713dc and https://github.com/kroxylicious/kroxylicious/pull/1366/commits/aa71624472fd4099269dfd03623b67737c15f604

PaulRMellor commented:
Thanks for the review comments @robobario 
Updates: https://github.com/kroxylicious/kroxylicious/pull/1366/commits/03499fbe6f924720b60029e41d5ef22b688903c8

k-wall commented on DEV_GUIDE.md:
```diff
@@ -3,37 +3,35 @@
-    * [Using the GitHub CI workflows against a fork](#using-the-github-ci-workflows-against-a-fork)
-  * [DCO Signoff](#dco-signoff)
-* [Deprecation Policy](#deprecation-policy)
+- [Development Guide for Kroxylicious](#development-guide-for-kroxylicious)
```
Is there a reason for the change? I don't object.

k-wall commented on docs/README.md:
```diff
@@ -1,18 +1,19 @@
-This folder contains the source files for managing the Kroxylicious documentation. The documentation is written in [AsciiDoc][AsciiDoc] and generated using 
- and it provides content to help you understand and set up Kroxylicious proxies.
+This folder contains the source files for managing the Kroxylicious documentation. 
+The documentation is written in [AsciiDoc][AsciiDoc] and provides content to help you understand and set up Kroxylicious proxies.
```
we've been generally referring to the component as simply "Kroxylicious" rather than "Kroxylicious proxy".
I think the existing ref was an exception rather than a rule.

```suggestion
The documentation is written in [AsciiDoc][AsciiDoc] and provides content to help you understand and set up Kroxylicious proxies.
```

k-wall commented on docs/assemblies/assembly-overview.adoc:
```diff
@@ -6,7 +6,7 @@
 
 [role="_abstract"]
-Kroxylicious Proxy is an Apache Kafka(R) protocol-aware ("Layer 7") proxy designed to enhance Kafka-based systems.
+Kroxylicious Proxy is an Apache Kafka protocol-aware ("Layer 7") proxy designed to enhance Kafka-based systems.
```
```suggestion
Kroxylicious is an Apache Kafka protocol-aware ("Layer 7") proxy designed to enhance Kafka-based systems.
```

k-wall commented on docs/assemblies/assembly-overview.adoc:
```diff
@@ -6,7 +6,7 @@
 = Kroxylicious Proxy overview
```
```suggestion
= Kroxylicious overview
```

k-wall commented on docs/assemblies/assembly-record-encryption-filter.adoc:
```diff
@@ -6,18 +6,21 @@
 
 [role="_abstract"]
-Streams for Apache Kafka Proxy's Record Encryption filter enhances the security of Kafka messages.
+Kroxylicious's Record Encryption filter enhances the security of Kafka messages.
```
Do we want to introduce a variable for Kroxylicious to facilitate the production of derivative works?

k-wall commented on docs/modules/con-custom-filters.adoc:
```diff
@@ -7,7 +7,8 @@
 Custom filters can be written in the Java programming language.
-Knowledge of the https://kafka.apache.org/protocol.html[Kafka protocol] is generally required to write a protocol filter.
+Kroxylicious supports Java 17.
+Knowledge of the https://kafka.apache.org/protocol.html[Kafka protocol^] is generally required to write a protocol filter.
```
```suggestion
Knowledge of the {kafka-protocol}[Kafka protocol^] is generally required to write a protocol filter.
```

k-wall commented on docs/modules/con-custom-filters.adoc:
```diff
@@ -23,9 +24,9 @@ You can find them {github}/tree/main/kroxylicious-sample[here] for a hands-on in
 == API docs
 
 Custom Protocol Filters are built by implementing interfaces supplied by the
```
We are referring to them as simple Filters.
```suggestion
Custom Filters are built by implementing interfaces supplied by the
```

k-wall commented on docs/modules/record-encryption/con-record-encryption-overview.adoc:
```diff
@@ -50,7 +50,7 @@ The filter decrypts records from fetch responses as follows:
 . Replaces the cipher record with a decrypted record.
 
-The filter uses an LRU (least recently used) strategy for decrypted records.
+The filter uses an LRU (least recently used) strategy for caching decrypted DEKS.
```
```suggestion
The filter uses an LRU (least recently used) strategy for caching decrypted DEKs.
```

k-wall commented on docs/modules/record-encryption/con-record-encryption-overview.adoc:
```diff
@@ -62,10 +62,10 @@ To support the filter, the KMS provides the following:
 
-WARNING: The KMS must be available during runtime.
-If the KMS is unavailable, production and consumption through the filter become impossible until KMS service is restored.
+WARNING: The KMS must be available during runtime. 
```
```suggestion
WARNING: The KMS must be available during runtime.
```

k-wall commented on docs/README.md:
```diff
@@ -1,18 +1,19 @@
- and it provides content to help you understand and set up Kroxylicious proxies.
+This folder contains the source files for managing the Kroxylicious documentation. 
+The documentation is written in [AsciiDoc][AsciiDoc] and provides content to help you understand and set up Kroxylicious proxies.
 
```
Let's mention that US English spells should be used.  Also is there a convention we want to adopt about the number of spaces used after a full-stop?  I tend to favour two spaces, but I really don't mind switching to one if that's what's prefered.

k-wall commented on docs/README.md:
````diff
@@ -37,11 +38,22 @@ A module contain a procedure (`proc-`), concepts (`con-`), or reference (`ref-`)
+mvn org.asciidoctor:asciidoctor-maven-plugin:process-asciidoc@convert-to-html
+```
+
+The HTML is output to `target/html/master.html`. 
````
the filename is `index.html`.

```suggestion
The HTML is output to `target/html/index.html`. 
```

k-wall reviewed (approved):
couple of nits and suggestions, but otherwise lgtm.

showuon commented:
@PaulRMellor , do you have time to address @k-wall 's comments above? And we think it's in a good shape to be merged, if no other comments.

PaulRMellor commented:
> @PaulRMellor , do you have time to address @k-wall 's comments above? And we think it's in a good shape to be merged, if no other comments.

Thanks @showuon -- I've addressed the comments now
